ocaml/lambda/matching.ml

```diff
@@ -129,65 +129,30 @@ let may_compat = MayCompat.compat
 
 and may_compats = MayCompat.compats
 
-(* The free variables in a guarded rhs can be precomputed to be used in an
-   optimization, or uncomputed, in which case the optimization is not applied.
-*)
-type guarded_free_variables =
-  | Precomputed of Ident.Set.t
-  | Uncomputed
-
-let map_guarded_free_variables ~f = function
-  | Precomputed free_variables -> Precomputed (f free_variables)
-  | Uncomputed -> Uncomputed
-
 type rhs =
-  | Guarded of
-      { patch_guarded: patch:lambda -> lambda
-      ; free_variables: guarded_free_variables }
-  (* Guarded rhs's must allow for fallthrough if the guard fails.
-
-     When translating a guarded rhs, the code to execute on fallthrough must
-     be "patched" in. Because the fallthrough code is translated after the
-     rhs is created, guarded actions rhs's a function [patch_guarded], which
-     generates a lambda term for the action from the fallthrough code.
-
-     One translation optimization requires us to check the free variables of an
-     rhs. When it is simple to do so, we precompute these free variables for use
-     in the optimization. Otherwise, we leave them uncomputed and do not apply
-     the optimization.
+  | Guarded of { action: lambda; exn_label: static_label }
```

Comment:
Can we name `action` something else? I'd like to avoid introducing "action" as a term. Maybe the name could suggest a stronger relationship with `arg_label`, something like:

```ocaml
{ rhs_with_fallthrough : lambda; fallthrough : static_label }
```

The comment could simply say "`rhs_with_fallthrough` raises exn `fallthrough` if the guard fails."